returnn/tf/engine.py

```diff
@@ -2285,15 +2285,18 @@ def search(self, dataset, do_eval=True, output_layer_names="output", output_file
         print("Given output %r has beam %s." % (output_layer, out_beam), file=log.v1)
         output_layer_beam_scores.append(output_layer.get_search_choices().beam_scores)
       out_beam_sizes.append(out_beam.beam_size if out_beam else None)
-      target_keys.append(output_layer.target or self.network.extern_data.default_target)
+      target_key = output_layer.target
+      if output_layer.output.sparse and not target_key:
+        # Use the default target key for sparse outputs
+        target_key = self.network.extern_data.default_target
+      target_keys.append(target_key)
```

Comment:
`target_key` can be `None` here. In that case, it should not be added to `target_keys`.

Reply:
Ah sorry, I realize that `target_keys` needs to match all the other lists...

Reply:
It probably makes sense to rewrite this part of the code a bit. Later in the code I still iterate over `range(num_targets)` and there I assume that the length of target_keys is equal to the number of output layers.

Reply:
Yea...